miden-crypto/src/hash/algebraic_sponge/mod.rs

```diff
@@ -49,40 +55,47 @@ const BINARY_CHUNK_SIZE: usize = 7;
 ///
 /// The constants are defined for tests only because the exponentiations in the code are unrolled
 /// for efficiency reasons.
-#[cfg(test)]
-const ALPHA: u64 = 7;
-#[cfg(test)]
-const INV_ALPHA: u64 = 10540996611094048183;
+#[cfg(all(test, feature = "std"))]
+pub(crate) const ALPHA: u64 = 7;
+#[cfg(all(test, feature = "std"))]
+pub(crate) const INV_ALPHA: u64 = 10540996611094048183;
 
 // ALGEBRAIC SPONGE
 // ================================================================================================
 
-pub(crate) trait AlgebraicSponge {
+// Note: This trait must remain `pub` (not `pub(crate)`) because it's re-exported
+// at the crate root (lib.rs) as part of the public API.
+pub trait AlgebraicSponge {
     fn apply_permutation(state: &mut [Felt; STATE_WIDTH]);
 
     /// Returns a hash of the provided field elements.
     fn hash_elements<E>(elements: &[E]) -> Word
     where
-        E: FieldElement<BaseField = Felt>,
+        E: BasedVectorSpace<Felt>,
     {
-        // convert the elements into a list of base field elements
-        let elements = E::slice_as_base_elements(elements);
+        // Count total number of base field elements without collecting
+        let total_len = elements
+            .iter()
+            .map(|elem| E::as_basis_coefficients_slice(elem).len())
+            .sum::<usize>();
```

Comment:
I'm assuming there is no function in P3 that can convert a slice of extension field elements into a slice of base field elements - right? If so, this may be a good function to introduce there. Could we make a note of this?